docs/pages/guides/onchain-indices.mdx

```diff
@@ -0,0 +1,412 @@
+
+   <CollapseCode>
+
+```solidity filename="MoveSystem.sol" copy showLineNumbers {11-14}
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.24;
+
+import { System } from "@latticexyz/world/src/System.sol";
+import { Position } from "../codegen/index.sol";
+
+contract MoveSystem is System {
+  function move(int32 x, int32 y) public {
+    address player = _msgSender();
+
+    bool positionIsEmpty = ?
+    require(positionIsEmpty, "another player is at this position");
+
+    Position.set(player, x, y);
+  }
+}
+```
+
+   </CollapseCode>
+
+In other words, there is no reverse mapping - we can read the position of a given player, but not which players are at each position. It's time to add an index!
+
+There are two main ways to implement this - you can **create and manage your own indices**, or **use a module**.
+
+### Creating your own indices
+
+#### `PlayersAtPosition`
+
+One way to know whether a position is occupied is to have a list of which players are there, and check if that list is empty.
+
+First, we declare an `PlayersAtPosition` table in our MUD config. The schema is essentially the reverse of `Position`; the key is an (x, y) coordinate and the value is an array of addresses currently at that position:
+
+   <CollapseCode>
+```ts filename="mud.config.ts" copy showLineNumbers {14-17}
+import { mudConfig } from "@latticexyz/world/register";
+
+export default mudConfig({
+  tables: {
+    Position: {
+      keySchema: {
+        player: "address",
+      },
+      valueSchema: {
+        x: "int32",
+        y: "int32",
+      },
+    },
+    PlayersAtPosition: {
+      keySchema: { x: "int32", y: "int32" },
+      valueSchema: { players: "address[]" },
+    },
+  },
+});
+```
+
+  </CollapseCode>
+
+With our own indices, we need to be careful that they are maintained whenever the relevant tables are updated. Otherwise, they will not reflect current store data (for example, players in a `PlayersInPosition` list that don't actually have that `Position` value).
+
+In this case, we need to add or remove addresses from `PlayersAtPosition` whenever a position is set. A good practice for this is to write a helper that atomically sets the table and index.
+
+We can use a `setPosition` helper function to set both `Position` and the index.
+
+```solidity filename="setPosition.sol" copy showLineNumbers
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.24;
+
+import { Position, PositionData, PlayersAtPosition } from "../codegen/index.sol";
+
+function filter(address[] memory arr, address element) pure returns (address[] memory) {
+  address[] memory filtered = new address[](arr.length);
+  uint256 filteredIndex;
+  for (uint256 i; i < arr.length; i++) {
+    if (arr[i] != element) {
+      filtered[filteredIndex] = arr[i];
+      filteredIndex++;
+    }
+  }
+
+  // In-place update the length of the array
+  // (Note: this does not update the free memory pointer)
+  assembly {
+    mstore(filtered, filteredIndex)
+  }
```

Comment:
I'd add `/// @solidity memory-safe-assembly` to this assembly block, though it wouldn't matter much either way